File: Z_xiaonei_class/xiaonei.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
'''
	BY PopA
	此程序由python 2.6.5 以及 ubuntu 10.04LTS下调试通过
	参考了众多的代码。感谢前驱者无限的帮助
	版权所有，任何人可以修改，复制，传播，但必须注明原作者
	欢迎交流
	contact:
		QQ:447521915
		Email:popacai@gmail.com
'''
import urllib,urllib2,cookielib,re,time
import sys
def getTicket(html):
	a = html.find("get_check:'")
	a = a + 11
	b = html.find("'",a)
	string = html[a:b]	
	return string
class xiaonei():
	def __init__(self,email,password):
		self.url = 'http://www.renren.com/'
		self.email = email
		self.password = password
		self.cookie = cookielib.LWPCookieJar() 
		self.opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.cookie))
		self.friendIdList = []
	
	def login(self):
		url = self.url+'PLogin.do/'
		params = urllib.urlencode({'email':self.email,'password':self.password})
		response = self.opener.open(urllib2.Request(url,params))
		url = self.url+'home'
		if response.geturl()==url:
			print 'Login Succesfully!'
			self.cookie.save('xiaonei.cookie')  
			self.ticket = ""
			response = self.opener.open('http://www.renren.com/home').read()
			self.ticket = getTicket(response)
			print self.ticket
		else:
			print 'Login Failed'
	def changeState(self,newStatus):
		ticket = self.ticket
		s = newStatus
		status = {"isAtHome":"1"}
		status['c'] = s
		status['raw'] = s
		status['publisher_form_ticket'] = ticket
		status['requestToken'] = ticket
		response = self.opener.open('http://status.renren.com/doing/update.do?',urllib.urlencode(status))
		print 'ok', response.read()
		
	def leaveMsg(self,ID,msg = 'hello'):
		ticket = self.ticket
		status = {"body":msg}
		status ['id'] = ID 
		status ['cc'] = ID
		status ['cccc='] = ''
		status ['tsc'] = ''
		status ['profilever'] = 2008
		status ['headUrl'] = ''
		status ['largeUrl'] = ''
		status ['requestToken'] = ticket
		status ['only_to_me'] = '0'
		status ['color'] = ''
		status ['ref'] = r'http://www.renren.com/profile.do'
		status ['mode'] = ''
		status ['requestToken'] = ticket
		response = self.opener.open('http://gossip.renren.com/gossip.do',urllib.urlencode(status)).read()
		if (response.find("visiter") != -1):
			print 'ok'
		else:
			print 'error, fail to sent'
	def getFriendList(self):
		url = 'http://friend.renren.com/myfriendlistx.do'
		response = self.opener.open(urllib2.Request(url))
		html = response.read()
		friend = str(re.search('friends=/[{.*}/]',html).group())
		friendId = re.findall('"id":/d+',friend)
		for id in friendId:
			self.friendIdList.append(id.lstrip('"id":'))
			
	def browseFriend(self):
		cnt=0
		for friendId in self.friendIdList:
			cnt+=1
			url='http://www.renren.com/profile.do?id='+'%s'%(friendId)
			response = self.opener.open(urllib2.Request(url))
			html = response.read()
#			f = open("1.html","w")
#			f.write(html)
			x = html.find('<title>')
			y = html.find('</title>')
			x += 26
			s = html[x:y]
			print "%d/%d %s" % (cnt,len(self.friendIdList),s)
	def sayHello(self, word):
		cnt = 0
		for friendId in self.friendIdList:
			cnt += 1
			self.leaveMsg(friendId,word)
			print "%d/%d %s" % (cnt,len(self.friendIdList),friendId)
			
def main():
	if len(sys.argv) < 2:
		print 'usage: email@xxx.com passwd'
		exit()	
	app = xiaonei(sys.argv[1],sys.argv[2])
	app.login()
	app.getFriendList()
	
	app.changeState('hello!')
#leavemsg

	Id = ''
	string = 'UTF-8数据'
	i = 0
	while (i < len(string)):
		if string[i] > '~':
			tstr = string[i:i+3]
			print tstr
			app.leaveMsg(Id,tstr)
			i+=3
		else:
			tstr = string[i]
			print tstr
			app.leaveMsg(Id,tstr)
			i+=1
		time.sleep(5)	

	app.browseFriend() #100名后需要验证码=.=
main()

File: Z_xiaonei_class/xiaonei.php
<?php
/* publish in GPLv3
 * by PopA
 * contact:
 *		mail: popacai@gmail.com
 *		qq: 447521915
 * enjoy!
 */

class Xiaonei
{
	var $cookie_jar;
	var $get_check;

	function login($username, $password) {
		$renren_username = $username;
		$renren_password = $password;
	
		$this->cookie_jar	= tempnam('./tmp','cookie');
		$ch = curl_init();
		curl_setopt($ch,CURLOPT_COOKIEJAR,$this->cookie_jar);
		curl_setopt($ch,CURLOPT_URL,"http://passport.renren.com/PLogin.do");
		curl_setopt($ch,CURLOPT_POST,TRUE);
		curl_setopt($ch,CURLOPT_FOLLOWLOCATION,TRUE);
		curl_setopt($ch,CURLOPT_POSTFIELDS,
			'email='.urlencode($renren_username).
			'&password='.urlencode($renren_password).
			'&autoLogin=true'.
			'&origURL=http%3A%2F%2Fwww.renren.com%2FHome.do&domain=renren.com');
		curl_setopt($ch,CURLOPT_RETURNTRANSFER,TRUE);
		$str = curl_exec($ch);
		$pattern = "/get_check:'([^']+)'/";
		preg_match($pattern,$str,$matches);
		$this->get_check = $matches[1];	
		curl_close($ch);
	}


	function changestate($item)
	{
		$ch1 = curl_init();
		curl_setopt($ch1,CURLOPT_URL,"http://status.renren.com/doing/update.do");
		curl_setopt($ch1,CURLOPT_COOKIEFILE,$this->cookie_jar);
		curl_setopt($ch1,CURLOPT_POST,TRUE);
		curl_setopt($ch1,CURLOPT_POSTFIELDS,'c='.urlencode($item).'&raw='.urlencode($item).'&isAtHome=1&publisher_form_ticket='.$this->get_check.'&requestToken='.$this->get_check);
		curl_setopt($ch1,CURLOPT_FOLLOWLOCATION,TRUE);
		curl_setopt($ch1,CURLOPT_REFERER,'http://status.renren.com/ajaxproxy.htm');
		curl_setopt($ch1,CURLOPT_RETURNTRANSFER,TRUE);
													 
		$ret = curl_exec($ch1);
		curl_close($ch1);
		$ret = json_decode($ret,TRUE);
		echo $ret['msg']."\n";
	}
	function leaveMsg($ID=0, $msg)
	{
		$ch1 = curl_init();
		curl_setopt($ch1,CURLOPT_URL,"http://gossip.renren.com/gossip.do");
		curl_setopt($ch1,CURLOPT_COOKIEFILE,$this->cookie_jar);
		curl_setopt($ch1,CURLOPT_POST,TRUE);
		curl_setopt($ch1,CURLOPT_POSTFIELDS,
			'body='.urlencode($msg).
			'&id='.urlencode($ID).
			'&cc='.urlencode($ID).
			'&cccc='.
			'&tsc='.
			'&profilever=2008'.
			'&headUrl='.
			'&lageUrl='.
			'&requestToken='.urlencode($this->get_check).
			'&only_to_me=0'.
			'&color='.
			'&ref=http://www.renren.com/profile.do'.
			'&mode=');
		curl_setopt($ch1,CURLOPT_FOLLOWLOCATION,TRUE);
		curl_setopt($ch1,CURLOPT_RETURNTRANSFER,TRUE);
													 
		$ret = curl_exec($ch1);
		curl_close($ch1);
		$ret = json_decode($ret,TRUE);
		echo $ret['visiter']."\n";
	}
	function send_msg()
	{
		$msg='
			';
	}
}

$xiaonei = new Xiaonei;
$xiaonei->login("popacai@gmail.com","(passwd)");
?>
